# Algorithms - Recursion

Recursion is a way to decompose an algorithm into a **base-case** and a way to roll-up that base case into a result. You generally see this in the form of a **function that calls itself**.

## Recursion

### Example: Factorial

A factorial in mathematics is a function on integers denoted by the $!$ symbol. It means we multiply the number by the product of all smaller natural numbers:

$$n! = n * (n-1) * (n-2) * ... * 1,\ n \in \mathbb{N}$$

In code this translates to a clean loop:

In [1]:
# factorial example
def factorial_naive(n):
    if n <= 1:
        return n
    res = 1
    for i in range(1,n+1):
        res *= i
    return res

factorial_naive(5)

120

In [2]:
# factorial example w/recursion

def factorial_recursive(n):
    #Base case
    if n == 1:
        return 1
    #Roll up the base case recursively
    else:
        return n * factorial_recursive(n-1)
    
factorial_recursive(5)

120

### Break it down

5 * factorial(4) 5 * 24 = 120

4 * factorial(3) 4 * 6 = 24

3 * factorial(2) 3 * 2 = 6

2 * factorial(1) (base case - returns 1) 2 * 1 = 2

### Exercise: Fibonnaci numbers

The [Fibonnaci numbers](https://en.wikipedia.org/wiki/Fibonacci_number) are a natural sequence of numbers representing growth. The **Fibonnaci function** $F_n$ is defined as follows:

$$F_0 = 0,\ F_1 = 1$$

$$F_n = F_{n-1} + F_{n-2}$$

The sequence looks like 

$$0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, ...$$

And it appears all over nature! Some flower petals shape and pineapples follow these numbers. 

**exercise:** Write a function `fibonacci(n)` which computes the n-th fibonacci number

In [3]:
def fib(n):
    """
    Takes a number n and returns
    it's Fibonnaci equivalent
    """
    if n == 0:
        return 0
    elif n ==1:
        return 1
    else:
        return fib(n-1) + fib(n-2)
     
fib(3)


2

## Divide-and-Conquer Algorithms

(section 2.3.1 in CLRS)

Often, algorithm problems seem intimidating because of their complexity.

The **divide and conquer** paradigm is a way to design solutions that manage this complexity into an elegant design.

The idea is to 

1. Break the problem into small subproblems that are similar to the original problem but trivial to solve
2. Solve the subproblems recursively
3. Combine these solutions to create a solution to the original problem.

## MergeSort

`merge_sort` is a sorting algorithm which is a huge improvement (in terms of **big-O**) over our previous `selection_sort` algorithm.

The idea for `merge_sort` is to:

- **Divide:** Divide the n-element sequence to be sorted into two subsequences of n=2 elements each.
- **Conquer:** Sort the two subsequences recursively using merge sort. 
- **Combine:** Merge the two sorted subsequences to produce the sorted answer.

The pseudocode for mergesort is as follows:

```
MergeSort(A):
size = len(A)
If size > l
     1. Find the middle point to divide the array into two halves:  
             middle m = (l+r)/2
     2. Call mergeSort for first half:   
             A_1 = mergeSort(A[:m])
     3. Call mergeSort for second half:
             A_2 = mergeSort(A[m:])
     4. Merge the two halves sorted in step 2 and 3:
             A =  merge(A_1, A_2)
```

Here, the `merge` method takes in two *sorted* arrays, and merges them together such that the result is sorted.

The actual `merge_sort` code is exquisitely simple:

In [4]:
# merge_sort

def merge_sort(arr):
    size = len(arr)
    if size > 1:
        m = size // 2
        left = merge_sort(arr[:m])
        right = merge_sort(arr[m:])
        return merge(left,right)
    else:
        return arr

We just need to implement the `merge` function, which is a little less clean:

In [5]:
def merge(left, right):
    res = []
    #zip in together left and right parts
    while left and right:
        if left[0] < right[0]:
            res.append(left[0]) #append to result
            left.pop(0) #remove val we appended
        else:
            res.append(right[0])
            right.pop(0)
            
    #Copy in remaining elements of left and right (if any)
    for i in left:
        res.append(i)
    for i in right:
        res.append(i)
    return res

A = [111,4,3,22,5,44.4,66.6,777]

merge_sort(A)

[3, 4, 5, 22, 44.4, 66.6, 111, 777]

You have just proven that merge sort has a time complexity of around n log(n), dropping some constants and insignificant terms.

### Merge Sort Analysis

We can do a bit of math to see what the **Big-O** of merge sort is. 

You won't have to do math like this yourself, but walking through it will illustrate the analysis logic. Let's divide the Big-O of the 3 parts:

- **Divide:** The divide step just computes the middle of the subarray, which takes constant time. Thus it's $O(1)$
- **Conquer:** We recursively solve two subproblems, each of size $n/2$, which contributes $2 * O(n/2)$.
- **Combine:** The `merge` procedure on two n-element subarrays takes time $O(n)$

Then the runtime for any "level" in the recursion is:

$$O(n) = 2 * O(n/2) + n$$

This builds a tree of function calls where each tree divides $n$ by half.

![](../assets/Merge-Sort-Recursion.png)

The number of divisions by 2 you can do on any number is $log_2(n)$. So the height of the tree is also $log_2(n)$ which calls for a total of 

$$O(n*log_2(n))$$

Note that $n*log_2(n) < n^2$ for large values of $n$